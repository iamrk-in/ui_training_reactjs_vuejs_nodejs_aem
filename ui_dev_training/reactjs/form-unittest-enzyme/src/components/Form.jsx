import React from "react";
import TextInput from "./TextInput";
import CheckBox from "./CheckBox";
import Button from "./Button";

class Form extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            firstname: '', 
            lastname: '', 
            subscribed: false
        }
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange(event) {
        const target = event.target;
        const name = target.name;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        this.setState({[name]: value}, () => {
            console.log(this.state[name])
        })
    }

    handleSubmit() {
        var subscribed = this.state.subscribed ? 'yes' : 'no';
        alert(`firstname: ${this.state.firstname}`);
    }


    render() {
        // let languages = ['english', 'spanish', 'germany', 'japanes'];
        return(
            <form name="frm" onSubmit={this.handleSubmit}>
                <TextInput labelFor="firstname" label="firstname" className="form-control" type="text" name="firstname" id="firstname" value={this.state.firstname} handleChange={this.handleChange} />

                <TextInput labelFor="lastname" label="lastname" className="form-control" type="text" name="lastname" id="lastname" value={this.state.lastname} handleChange={this.handleChange} />

                <CheckBox labelFor="subscribed" label="subscribed" handleChange={this.handleChange} Checked={this.state.subscribed} className="form-check mb-3 text-right" />
                
                <Button type="submit" className="btn btn-primary"> submit </Button>
            </form>
        )
    }
}


export default Form;