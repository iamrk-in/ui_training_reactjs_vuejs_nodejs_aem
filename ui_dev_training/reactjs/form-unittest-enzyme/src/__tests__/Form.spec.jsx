import React from "react";
import renderer from "react-test-renderer";

import { mount } from "enzyme";
import enzymeConfig from "../../enzymeConfig";
import Form from "../components/Form";

describe("Form", function() {
    
    it("should render correctly", function() {
        const tree = renderer.create(<Form />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    
    it("should capture firstname correctly onChange", function() {
        const component = mount(<Form />);
        const input = component.find('input').at(0);
        input.instance().value = 'hello';
        input.simulate('change');
        expect(component.state().firstname).toEqual('hello');
    });


    it("should capture lastname correctly onChange", function() {
        const component = mount(<Form />);
        const input = component.find('input').at(1);
        input.instance().value = 'world';
        input.simulate('change');
        expect(component.state().lastname).toEqual('world');
    });

    it('should capture checkbox ticked correctly onChange', function() {
        const component = mount(<Form />);
        const input = component.find('input').at(2);
        input.instance().checked = true;
        input.simulate('change');
        expect(component.state().subscribed).toEqual(true);
    });

    it('should call alert() when submit button is clicked', function() {
        const state = {firstname:'hello', lastname:'world', subscribed: true};
        const expectedArg = "firstname: hello, lastname: world, subscribed: yes";
        const component = mount(<Form />);
        window.alert = jest.fn();
        component.find('form').simulate('submit');
        expect(window.alert).toHaveBeenCalledWith(expectedArg);
    });

});