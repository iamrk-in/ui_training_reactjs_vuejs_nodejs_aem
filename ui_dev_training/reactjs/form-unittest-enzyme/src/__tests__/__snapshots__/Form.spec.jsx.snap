// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<form
  name="frm"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      htmlFor="firstname"
    >
      firstname
    </label>
    <input
      className="form-control"
      id="firstname"
      name="firstname"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="firstname"
    >
      lastname
    </label>
    <input
      className="form-control"
      id="lastname"
      name="lastname"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-check mb-3 text-right"
  >
    <input
      className="form-check-input"
      id="subscribed"
      name="subscribed"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="form-check-label"
      htmlFor="subscribed"
    >
      subscribed
    </label>
  </div>
  <button
    className="btn btn-primary"
    type="submit"
  >
     submit 
  </button>
</form>
`;
